### PR TITLE
Refactor FXIOS-22527 - UI Tests Added waitAndTap() for better maintenance

### PR DESCRIPTION
**Pull request by hemche**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10282)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/22527)

## :bulb: Description

This issue is regarding the UITests at https://github.com/mozilla-mobile/firefox-ios/tree/main/firefox-ios/firefox-ios-tests/Tests
Currently, we are using the below code whenever we need to wait for an element's existence and then tap on it.

```
mozWaitForElementToExist(app.buttons["XXX"])
app.buttons["XXX"].tap()
```
There are many places where we are using the above two lines. It would be beneficial in terms of maintenance to have a function like `waitAndTap()` within the `XCUIElement`. So that in the future if we need to improve the wait method or change it due to syntax changes from Apple XCUITest side, it will be a change at one place instead of changing it in each and every test.

Updated al the tests wherever wait and then tap was being used.
Ran all the smoke XCUI tests locally and they passed. 1 failing test in smoke suite is not related to the change i did.

Screenshot 2024-10-13 at 10 41 47 PM
Screenshot 2024-10-13 at 9 59 02 PM

Screenshot 2024-10-13 at 10 16 47 PM
Screenshot 2024-10-13 at 10 30 45 PM

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-10282)

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by isabelrios** [comment]
@hemche There are swiftlint errors in this PR, please check them here: https://github.com/mozilla-mobile/firefox-ios/runs/31488235539

**Comment by hemche** [comment]
> @hemche There are swiftlint errors in this PR, please check them here: https://github.com/mozilla-mobile/firefox-ios/runs/31488235539

Thanks for the catch. Will update my linter settings for future PRs. Should be fixed in [5ee6d08](https://github.com/mozilla-mobile/firefox-ios/pull/22528/commits/5ee6d087587e28ffad27cc60b8a7235c42f59dd6)

**Comment by hemche** [comment]
This 1 failing test is not part of my refactoring.
This test is passing locally when i run. So, let me know if there is something else i need to fix before this PR can be reviewed.
Ofcourse whenever you have time. No rush. @isabelrios @jjSDET 
https://github.com/mozilla-mobile/firefox-ios/pull/22528/checks?check_run_id=31561753904

**Comment by hemche** [comment]
> This pull request has conflicts when rebasing. Could you fix it @hemche? 🙏

Rebase conflicts resolved in [cfc400b](https://github.com/mozilla-mobile/firefox-ios/pull/22528/commits/cfc400b41ebe8c132018790345c8d106821a2fec)

**Comment by clarmso** [comment]
Smoke tests look good: https://app.bitrise.io/app/6c06d3a40422d10f/pipelines/e12a5a6f-d309-4a7b-9fb9-a10d82445c93

L10n tests also look good. I would approve 👍🏼 this PR.

**Review by clarmso — APPROVED** [review]

File: firefox-ios/firefox-ios-tests/Tests/L10nSnapshotTests/L10nBaseSnapshotTests.swift [hunk 1]
```diff
@@ -131,4 +131,11 @@ extension XCUIElement {
     func tapOnApp() {
         coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.5)).tap()
     }
+
+    func waitAndTap(timeout: TimeInterval? = TIMEOUT) {
+        L10nBaseSnapshotTests().mozWaitForElementToExist(self, timeout: timeout)
+        if self.exists {
```
Comment (mdotb-moz) [inline firefox-ios/firefox-ios-tests/Tests/L10nSnapshotTests/L10nBaseSnapshotTests.swift:137]: Can we add some code to improve the readability and check if the condition isn't met, terminate early the method? I think we can improve this with the code

`
guard self.exists else {
  XCTFail("Element not found whithin \(timeout ?? 0) seconds")
            return 
}
// Tap the element if it exists
self.tap()
`

or with an if else, like this:
`if self.exists {
        self.tap()
    } else {
        XCTFail("Element not found within \(timeout ?? 0) seconds")
    }`

What do you think?

File: firefox-ios/firefox-ios-tests/Tests/XCUITests/BaseTestCase.swift [hunk 2]
```diff
@@ -512,6 +511,13 @@ extension XCUIElement {
         let visibleScreenFrame = getVisibleScreenFrame(app: app)
         return self.exists && isPartiallyIncluded(rectangleArea: visibleScreenFrame, rectangleToBeIncluded: self.frame)
     }
+    /// Waits for the UI element and then taps it if it exists.
+    func waitAndTap(timeout: TimeInterval? = TIMEOUT) {
+        BaseTestCase().mozWaitForElementToExist(self, timeout: timeout)
+        if self.exists {
```
Comment (mdotb-moz) [inline firefox-ios/firefox-ios-tests/Tests/XCUITests/BaseTestCase.swift:517]: same comment I did before

**Review by mdotb-moz — COMMENTED** [review]
I like the idea to reduce the code. I add a couple of comments to improve the code for readability.

File: firefox-ios/firefox-ios-tests/Tests/L10nSnapshotTests/L10nBaseSnapshotTests.swift (hunk 1 above)
Comment (hemche) [reply firefox-ios/firefox-ios-tests/Tests/L10nSnapshotTests/L10nBaseSnapshotTests.swift]: Updated `mozWaitForElementToExist` function to use `guard`. Thanks for the tip 🙏🏽 
Fixed here: https://github.com/mozilla-mobile/firefox-ios/pull/22528/commits/a50c8636d5aaee960a6340cd482fff64644695fc

File: firefox-ios/firefox-ios-tests/Tests/XCUITests/BaseTestCase.swift (hunk 2 above)
Comment (hemche) [reply firefox-ios/firefox-ios-tests/Tests/XCUITests/BaseTestCase.swift]: Updated `mozWaitForElementToExist` function to use `guard`. Thanks for the tip 🙏🏽 
Fixed here: https://github.com/mozilla-mobile/firefox-ios/pull/22528/commits/a50c8636d5aaee960a6340cd482fff64644695fc

File: firefox-ios/firefox-ios-tests/Tests/XCUITests/ActivityStreamTest.swift [hunk 3]
```diff
@@ -246,8 +244,7 @@ class ActivityStreamTest: BaseTestCase {
     }
 
     private func checkNumberOfExpectedTopSites(numberOfExpectedTopSites: Int) {
-        mozWaitForElementToExist(app.cells[TopSites.itemCell])
-        mozWaitForElementToExist(app.cells[TopSites.itemCell])
+        mozWaitForElementToExist(app.cells[TopSites.itemCell], timeout: 20)
```
Comment (clarmso) [inline firefox-ios/firefox-ios-tests/Tests/XCUITests/ActivityStreamTest.swift:247]: The default timeout is 20 so the `timeout:` argument isn't needed.

Comment (hemche) [reply firefox-ios/firefox-ios-tests/Tests/XCUITests/ActivityStreamTest.swift]: Fixed in faf7ad1bbd4eddfce845f870d8250af73ee0fd39

**Pull request merged**